### PR TITLE
`Communication`: Remove hidden conversations from favorites

### LINKED ISSUE #9349: `Communication`: Hidden favorite channels still visible in favorites
### Describe the bug

When a channel is added to favorites and then hidden, it still appears in the favorites section. Expected behavior is that hidden channels should only be visible in the hidden section, not in favorites.

### To Reproduce

1. Go to the Communication section of a course.
2. Add a channel to favorites by clicking the heart-shaped favorite button.
3. Notice that it is now in the favorites section. Click on its dropdown menu and select the hide option.
4. Notice that it is still in the favorites section and has not been added to the hidden section.

### Expected behavior

When a channel is hidden, it should only appear in the hidden section, and if it is in favorites, it should be removed from there.

### Screenshots

_No response_

### Which version of Artemis are you seeing the problem on?

7.5.3

### What browsers are you seeing the problem on?

Chrome

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by asliayk**
#### General

- [X] This is a small issue that I tested locally.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [X] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [X] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).

### Motivation and Context

- Hiding a favorite conversation does not remove it from the favorites list, so it remains unhidden and is not added to the hidden list.
- Favoriting a hidden conversation does not change its hidden state. Although it is added to the favorites list, it still displays a "show" button instead of "hide".
(Closes #9349)

### Description

- If the conversation is in favorites and is being hidden, it is added to the hidden list and also removed from the favorites list.
- If the conversation is hidden and is being favorited, it is added to the favorites list and also removed from the hidden list.

### Steps for Testing

Prerequisites:
- 1 Instructor/Student

1. Log in to Artemis
2. Navigate to Communication section of a course
3. Add a conversation to favorites
4. Click on hide button to add it to the hidden list
5. Notice that it is added to the hidden list and also removed from favorites.
6. Add that channel to the favorites again.
7. Notice that it is removed from the hidden list and added to the favorites list.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| conversation-options.component.ts | 91.83% | ✅ ❌ |

## Summary by CodeRabbit

- **New Features**
	- Enhanced conversation options to allow users to hide conversations while automatically managing favorite status.
	- Users can now favorite hidden conversations, which will unhide them.

- **Bug Fixes**
	- Improved logic for handling conversation visibility and favorite status to ensure consistent behavior.

- **Tests**
	- Added new test cases to ensure functionality for hiding and favoriting conversations is working as intended.

### CODE REVIEW

**Review by SimonEntholzer — APPROVED** [review]
Tested on Ts4 and works as described

**Review by PaRangger — APPROVED** [review]
Tested on TS5. Code LGTM 👍

**Review by sachmii — APPROVED** [review]
Tested on TS5, worked as described.

File: src/main/webapp/app/shared/sidebar/conversation-options/conversation-options.component.ts [hunk 1]
```diff
@@ -79,11 +79,33 @@ export class ConversationOptionsComponent implements OnInit, OnDestroy {
 
     onHiddenClicked(event: MouseEvent) {
         event.stopPropagation();
+        if (!this.course.id || !this.conversation.id) return;
+
+        // If the conversation is in favorites and is being hidden, it is also removed from favorites
```
Comment (az108) [inline src/main/webapp/app/shared/sidebar/conversation-options/conversation-options.component.ts:84]: In my opinion we can remove this comment as it is not really describing something complex in the method

File: src/main/webapp/app/shared/sidebar/conversation-options/conversation-options.component.ts [hunk 2]
```diff
@@ -79,11 +79,33 @@ export class ConversationOptionsComponent implements OnInit, OnDestroy {
...
+            });
+        }
         this.hide$.next(!this.conversation.isHidden);
     }
 
     onFavoriteClicked($event: MouseEvent) {
         $event.stopPropagation();
+        if (!this.course.id || !this.conversation.id) return;
+
+        // If the conversation is hidden and is being favorited, it is also removed from the hidden list
```
Comment (az108) [inline src/main/webapp/app/shared/sidebar/conversation-options/conversation-options.component.ts:100]: same here

**Review by az108 — COMMENTED** [review]
Overall the code lgtm, can we consider this change?

**Review by HawKhiem — DISMISSED** [review]
Tested on TS3. Code works as expected
Screenshot 2024-10-16 180306
Screenshot 2024-10-16 180315

**Review by HawKhiem — APPROVED** [review]
Screenshot 2024-10-17 200300
Screenshot 2024-10-17 200313

Tested on TS1. Still works as described

**Review by PaRangger — APPROVED** [review]
Re-tested on TS1. Still works as expected 👍

**Review by az108 — APPROVED** [review]
Thanks for the changes, Code 👍

**Review by sachmii — APPROVED** [review]
Tested on TS3, still works as described

**Review by krusche — APPROVED** [review]
Code looks good 👍

**Pull request merged**